Add config flow to HLK-SW16

jameshilliard opened the pull request:
## Breaking change
The hlk-sw16 entity ids will change upon update.  The old entities will need to be removed manually.

## Proposed change
This change converts the hlk-sw16 component to use config flow


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/14140

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jameshilliard committed:
Add config flow to HLK-SW16

frenck commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,134 +2,133 @@
-    vol.Any(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, "a", "b", "c", "d", "e", "f"), vol.Coerce(str)
-)
-
-CONFIG_SCHEMA = vol.Schema(
```
This instantly makes the YAML configuration invalid and can potentially break existing setups.
Can we import existing configuration?

jameshilliard replied:
I looked into that but it seemed quite difficult to migrate.

jameshilliard replied:
Is it fine to leave migrations unsupported? I'd really prefer not to support migrations for this device due to it being difficult to map the old config format to the new one. Due to the type of device I don't think it will be that big an issue for users to manually migrate.

bdraco replied:
I added a code suggestion to get you part of the way there.

bdraco commented on homeassistant/components/hlk_sw16/switch.py:
```diff
@@ -1,30 +1,29 @@
-        device = SW16Switch(device_name, device_port, device_id, device_client)
+    for i in range(16):
+        device_port = f"{i:01x}"
+        device = SW16Switch(device_port, address, device_client)
```
If you want something unique here, pass `entry.entry_id` to `SW16Switch`

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,134 +2,133 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return f"{self._address}:{self._device_port}"
```
```suggestion
        return f"{self._entry_id}_{self._device_port}"
```

jameshilliard committed:
Use entry_id for unique_id

jameshilliard committed:
Add options update capability

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
-    extra=vol.ALLOW_EXTRA,
-)
+async def async_setup(hass, config):
+    """Component setup, do nothing."""
```
```suggestion
    """Component setup, do nothing."""
    hass.data.setdefault(DOMAIN, {})
```

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
-                reconnect_interval=DEFAULT_RECONNECT_INTERVAL,
-                keep_alive_interval=DEFAULT_KEEP_ALIVE_INTERVAL,
-            )
+    if hass.data.get(DATA_DEVICE_REGISTER, None) is None:
+        hass.data[DATA_DEVICE_REGISTER] = {}
+    if hass.data.get(DATA_DEVICE_LISTENER, None) is None:
+        hass.data[DATA_DEVICE_LISTENER] = {}
```
```suggestion
    if hass.data.get(DATA_DEVICE_REGISTER, None) is None:
        hass.data[DATA_DEVICE_REGISTER] = {}
    if hass.data.get(DATA_DEVICE_LISTENER, None) is None:
        hass.data[DATA_DEVICE_LISTENER] = {}
```

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
+    port = entry.data[CONF_PORT]
 
-            hass.data[DATA_DEVICE_REGISTER][device] = client
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = entry
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = {}
```

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
+            keep_alive_interval=DEFAULT_KEEP_ALIVE_INTERVAL,
+        )
+
+        hass.data[DATA_DEVICE_REGISTER][entry.entry_id] = client
```
```suggestion
        hass.data[DOMAIN][entry.entry_id][DATA_DEVICE_REGISTER] = client
```

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
+        )
+
+        # handle shutdown of HLK-SW16 asyncio transport
+        hass.data[DATA_DEVICE_LISTENER][entry.entry_id] = hass.bus.async_listen_once(
```
```suggestion
        hass.data[DOMAIN][entry.entry_id][DATA_DEVICE_LISTENER] = hass.bus.async_listen_once(
```

bdraco replied:
Except you should probably move this to unload instead

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
+
+async def async_remove_entry(hass, entry):
+    """Stop client config entry."""
+    client = hass.data[DATA_DEVICE_REGISTER].pop(entry.entry_id)
```
All of this needs to happen on unload instead.  Remote is when we delete the entry.

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,142 @@
+    """Stop client config entry."""
+    client = hass.data[DATA_DEVICE_REGISTER].pop(entry.entry_id)
+    client.stop()
+    if not hass.data[DATA_DEVICE_REGISTER]:
```
Per suggestion above this is now under `hass.data[DOMAIN][entry.entry_id]...`

bdraco commented on homeassistant/components/hlk_sw16/switch.py:
```diff
@@ -1,30 +1,27 @@
-    switches = domain_config[0]
-    device_id = domain_config[1]
-    device_client = hass.data[DATA_DEVICE_REGISTER][device_id]
+    device_client = hass.data[DATA_DEVICE_REGISTER][entry.entry_id]
```
```suggestion
    device_client = hass.data[DOMAIN][entry.entry_id][DATA_DEVICE_REGISTER]
```

jameshilliard committed:
Refactor entry_id under domain

jameshilliard commented on homeassistant/components/hlk_sw16/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+
+        options = {
+            vol.Required(
+                CONF_NAME, default=self.config_entry.options.get(CONF_NAME),
```
any idea why `self.config_entry.options` is empty?

jameshilliard committed:
Remove name from config

jameshilliard committed:
Set options

bdraco commented on homeassistant/components/hlk_sw16/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        )
+
+
+class SW16OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a HLK-SW16 options flow."""
+
+    def __init__(self, config_entry):
+        """Initialize options flow."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Manage options."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        options = {
+            vol.Required(
+                CONF_HOST, default=self.config_entry.options.get(CONF_HOST),
+            ): str,
+            vol.Optional(
+                CONF_PORT,
+                default=self.config_entry.options.get(CONF_PORT, DEFAULT_PORT),
+            ): vol.Coerce(int),
+        }
+
+        return self.async_show_form(step_id="init", data_schema=vol.Schema(options))
```
```suggestion
```

Let's take the options flow out for the initial PR per the discussion on `#devs_core`

jameshilliard replied:
removed options flow for now

jameshilliard committed:
Remove options flow

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,125 @@
+
+async def async_remove_entry(hass, entry):
+    """Stop client config entry."""
+    if hass.data[DOMAIN][entry.entry_id]:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
```
Please move this to `async_unload_entry` as we don't want to leave this in memory when the config entry is unloaded.

You'll need to move `hass.data.setdefault(DOMAIN, {})` from `async_setup` back into `async_setup_entry` as well since deleting and readding will have called `hass.data.pop(DOMAIN)`

jameshilliard replied:
moved to `async_unload_entry`

bdraco commented on homeassistant/components/hlk_sw16/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        client = await asyncio.wait_for(client_aw, timeout=CONNECTION_TIMEOUT)
+    except asyncio.TimeoutError:
+        raise CannotConnect
+    else:
+        try:
+
+            def disconnect_callback():
+                if client.in_transaction:
+                    client.active_transaction.set_exception(CannotConnect)
+
+            client.disconnect_callback = disconnect_callback
+            await client.status()
+        except CannotConnect:
+            client.disconnect_callback = None
+            client.stop()
+            raise CannotConnect
+        else:
+            client.disconnect_callback = None
+            client.stop()
```
Can we delete the top `else:` and save some indentation here ?

jameshilliard replied:
removed

bdraco reviewed (changes requested):
We need tests for the config flow.  You can use `python3 -m script.scaffold config_flow` to get started.

You can verify coverage with:
` pytest --cov=homeassistant/components/hlk_sw16/ --cov-report term-missing -- tests/components/hlk_sw16/test_*.py`

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,125 @@
+async def async_setup(hass, config):
+    """Component setup, do nothing."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
```
If you leave `CONFIG_SCHEMA` in and add

```suggestion
    conf = config.get(DOMAIN)

    if not conf:
        return True

    hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": SOURCE_IMPORT},
            data=conf
        )
    )

    return True
```

Import should be as simple as this in `config_flow.py`

```
    async def async_step_import(self, user_input):
        """Handle import."""
        return await self.async_step_user(user_input)
```

jameshilliard replied:
How would this work when `CONFIG_SCHEMA` doesn't match the new `DATA_SCHEMA`?

bdraco replied:
Ignore `CONF_SWITCHES`

bdraco replied:
If you need a sample import test, you can borrow from `nut/test_config_flow.py`

jameshilliard replied:
So I can't seem to get this working properly, there's something messing with the exception handling for `validate_input` but I can't figure out the cause. This does not work as is since each yaml config for this device can contain multiple devices, so I have it looping though and importing all entries.

jameshilliard replied:
Here's more or less what I tried: https://github.com/jameshilliard/home-assistant/commit/3d5491f16bd51c8ce2c01860ee8d49f5da89715e

bdraco replied:
You are pretty close on this, you shouldn't have to make any changes to the config flow beyond 

```
    async def async_step_import(self, user_input):
        """Handle import."""
        return await self.async_step_user(user_input)
```

jameshilliard replied:
I tried that first, it didn't work.

bdraco replied:
I updated the suggestion above since the schema is a bit different for this integration.

jameshilliard replied:
Ok, I think that works, updated.

jameshilliard committed:
remove unneccesary else block from validate_input and move domain cleanup to async_unload_entry

jameshilliard committed:
Add tests and config import

jameshilliard committed:
Add back config schema

jameshilliard commented:
> We need tests for the config flow. You can use python3 -m script.scaffold config_flow to get started.

Added tests

jameshilliard committed:
Remove config import

jameshilliard commented:
I've gone ahead and removed the config import since it's looking to be way too difficult to get working properly.

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -2,139 +2,121 @@
+    """Unload a config entry."""
+    client = hass.data[DOMAIN][entry.entry_id].pop(DATA_DEVICE_REGISTER)
+    client.stop()
+    if hass.data[DOMAIN][entry.entry_id]:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
+    return await hass.config_entries.async_forward_entry_unload(entry, "switch")
```
```suggestion
    unload_ok = await hass.config_entries.async_forward_entry_unload(entry, "switch")
    
    if unload_ok:
       if hass.data[DOMAIN][entry.entry_id]:
           hass.data[DOMAIN].pop(entry.entry_id)
       if not hass.data[DOMAIN]:
           hass.data.pop(DOMAIN)

    return unload_ok
```

jameshilliard replied:
applied

jameshilliard committed:
Refactor unload

jameshilliard committed:
Add back config import

bdraco reviewed (changes requested):
Please update `.coveragerc` to include only files that lack test coverage.

```
diff --git a/.coveragerc b/.coveragerc
index c49ac6257d..6ea6440992 100644
--- a/.coveragerc
+++ b/.coveragerc
@@ -350,7 +350,8 @@ omit =
     homeassistant/components/hisense_aehw4a1/*
     homeassistant/components/hitron_coda/device_tracker.py
     homeassistant/components/hive/*
-    homeassistant/components/hlk_sw16/*
+    homeassistant/components/hlk_sw16/__init__.py
+    homeassistant/components/hlk_sw16/switch.py
     homeassistant/components/home_connect/*
     homeassistant/components/homematic/*
     homeassistant/components/homematic/climate.py
```

Also, please verify that there is 100% coverage on config_flow.py with
`pytest --cov=homeassistant/components/hlk_sw16/ --cov-report term-missing -- tests/components/hlk_sw16/test_*.py`
```
----------- coverage: platform linux, python 3.8.1-final-0 -----------
Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
homeassistant/components/hlk_sw16/config_flow.py      47     21    55%   30-63, 74
homeassistant/components/hlk_sw16/const.py             6      0   100%
homeassistant/components/hlk_sw16/errors.py            5      0   100%
--------------------------------------------------------------------------------
```

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,82 @@
+    )
+
+    with patch(
+        "homeassistant.components.hlk_sw16.config_flow.validate_input",
```
Please patch the external library here instead of patching `validate_input`

jameshilliard replied:
removed patch entirely

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,82 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.hlk_sw16.config_flow.validate_input",
```
Please patch the external library here instead of patching `validate_input`

jameshilliard replied:
removed patch entirely

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,82 @@
+    )
+
+    with patch(
+        "homeassistant.components.hlk_sw16.config_flow.validate_input",
```
Please patch the external library here instead of patching `validate_input`

jameshilliard replied:
removed patch entirely

jameshilliard committed:
Update coveragerc

jameshilliard committed:
Don't mock validate_input

jameshilliard committed:
Test duplicate configs

jameshilliard committed:
Add import test

jameshilliard commented:
> Also, please verify that there is 100% coverage on config_flow.py with

done
```python
collecting ... 
 tests/components/hlk_sw16/test_config_flow.py ✓✓✓✓                                                                                                                                                                    100% ██████████

---------- coverage: platform darwin, python 3.8.5-final-0 -----------
Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
homeassistant/components/hlk_sw16/config_flow.py      47      0   100%
homeassistant/components/hlk_sw16/const.py             6      0   100%
homeassistant/components/hlk_sw16/errors.py            5      0   100%
--------------------------------------------------------------------------------
TOTAL                                                 58      0   100%


Results (10.47s):
       4 passed
```

jameshilliard committed:
Use patch for timeout test

jameshilliard committed:
Use mock for testing timeout

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -23,45 +54,51 @@ async def test_form(hass):
 
     assert result2["type"] == "create_entry"
-    assert result2["title"] == "1.1.1.1:8080"
+    assert result2["title"] == "127.0.0.1:" + str(port)
```
```suggestion
    assert result2["title"] == f"127.0.0.1:{port}"
```
The Home Assistant standard is to use f-strings.

jameshilliard replied:
removed dynamic port usage

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,170 @@
+
+    port = free_port()
+
+    server = await asyncio.start_server(handle_hlk_sw16_status_read, "127.0.0.1", port)
```
Please mock out the library calls here, we don't want to do I/O in tests as this makes them unreliable.

jameshilliard replied:
created a mock client that simulates the library client responses

jameshilliard committed:
Use MockSW16Client for tests

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,180 @@
+
+    mock_hlk_sw16_connection = await create_mock_hlk_sw16_connection(False)
+
+    with patch(
+        "homeassistant.components.hlk_sw16.config_flow.create_hlk_sw16_connection",
+        return_value=mock_hlk_sw16_connection,
+    ):
```
```suggestion
    with patch(
        "homeassistant.components.hlk_sw16.config_flow.create_hlk_sw16_connection",
        return_value=mock_hlk_sw16_connection,
    ), patch(
        "homeassistant.components.hlk_sw16.async_setup", return_value=True
    ) as mock_setup, patch(
        "homeassistant.components.hlk_sw16.async_setup_entry", return_value=True,
    ) as mock_setup_entry::
```

bdraco replied:
We want to avoid setting up the integration in the test as well since that could do I/O

jameshilliard replied:
applied your suggested change, do I need to do this in any of the other tests as well?

bdraco replied:
Only if they actually get far enough to create the entry

jameshilliard replied:
Hmm, only other one I think that might do that would be `test_import()`, but I'm not sure, would that set up the integration or does it just import the config?

bdraco replied:
` assert result2["type"] == "create_entry"` its creating an entry so it needs setup mocked out as well

jameshilliard replied:
added

bdraco commented on tests/components/hlk_sw16/test_config_flow.py:
```diff
@@ -0,0 +1,180 @@
+        "host": "127.0.0.1",
+        "port": 8080,
+    }
+
```
```suggestion
    await hass.async_block_till_done()
    assert len(mock_setup.mock_calls) == 1
    assert len(mock_setup_entry.mock_calls) == 1

```

jameshilliard committed:
Check mock_calls count

bdraco commented on homeassistant/components/hlk_sw16/errors.py:
```diff
@@ -0,0 +1,18 @@
+class AlreadyConfigured(SW16Exception):
+    """HLK-SW16 is already configured."""
+
+
+class NameExists(SW16Exception):
+    """HLK-SW16 name already configured."""
+
+
```
```suggestion
```

Looks unused

jameshilliard replied:
removed

bdraco commented on homeassistant/components/hlk_sw16/strings.json:
```diff
@@ -0,0 +1,22 @@
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
+    }
+  },
+  "title": "Hi-Link HLK-SW16"
```
```suggestion
```

title will be pulled from the `manifest.json`

jameshilliard replied:
removed

jameshilliard committed:
Remove unused NameExists exception

jameshilliard committed:
Remove title from strings.json

jameshilliard committed:
Mock setup for import test

bdraco commented:
Looking good.  We do need a docs PR to tell users how to setup the integration now that is available though the UI as well as documenting the changes listed in the summary.

jameshilliard commented:
> We do need a docs PR to tell users how to setup the integration now that is available though the UI as well as documenting the changes listed in the summary.

I've added docs and updated the PR summary.

jameshilliard commented:
Should I be setting `PARALLEL_UPDATES` for this integration?

bdraco commented:
> Should I be setting `PARALLEL_UPDATES` for this integration?

        If parallel updates is set to 0, we skip the semaphore.
        If parallel updates is set to a number, we initialize the semaphore to that number.
        The default value for parallel requests is decided based on the first entity that is added to Home Assistant.
        It's 0 if the entity defines the async_update method, else it's 1.


Since there is no `async_update`, it should probably be added to `switch.py`

jameshilliard committed:
Set PARALLEL_UPDATES for switch

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -57,84 +54,113 @@
 
 async def async_setup(hass, config):
+    """Component setup, do nothing."""
+    hass.data.setdefault(DOMAIN, {})
```
```suggestion
```

jameshilliard replied:
removed

bdraco commented on homeassistant/components/hlk_sw16/__init__.py:
```diff
@@ -57,84 +54,113 @@
+
+async def async_setup_entry(hass, entry):
     """Set up the HLK-SW16 switch."""
     # Allow platform to specify function to register new unknown devices
```
```suggestion
        hass.data.setdefault(DOMAIN, {})   
```

Last thing, need to move this back to `async_setup_entry` since it can be poped when the last config entry is unloaded.

jameshilliard replied:
added/moved back

jameshilliard commented:
I've set it to 0 which seems to work fine(not sure I can tell much of a difference). Since the commands for entities are multiplexed already by the library through the same connection I don't think there's any way home assistant could overload the device.

jameshilliard committed:
Move hass.data.setdefault(DOMAIN, {}) to async_setup_entry

bdraco reviewed (approved):
Nice work 👍 

Please retest to make sure everything is still working.

jameshilliard commented:
> Please retest to make sure everything is still working.

Everything seems to all work good still.

MartinHjelmare commented on homeassistant/components/hlk_sw16/switch.py:
```diff
@@ -1,30 +1,30 @@
 
-_LOGGER = logging.getLogger(__name__)
 
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```
This can be removed now.

jameshilliard replied:
Fixed in #38648.

MartinHjelmare reviewed (commented):
Nice!
